pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,473 @@
+		oldUIDToRef[string(old[i].UID)] = old[i]
+	}
+	oldUIDSet := sets.StringKeySet(oldUIDToRef)
+	newUIDToRef := make(map[string]metatypes.OwnerReference)
+	for i := 0; i < len(new); i++ {
+		newUIDToRef[string(new[i].UID)] = new[i]
+	}
+	newUIDSet := sets.StringKeySet(newUIDToRef)
+
+	addedUID := newUIDSet.Difference(oldUIDSet)
+	removedUID := oldUIDSet.Difference(newUIDSet)
+
+	for uid, _ := range addedUID {
+		added = append(added, newUIDToRef[uid])
+	}
+	for uid, _ := range removedUID {
+		removed = append(removed, oldUIDToRef[uid])
+	}
+	return added, removed
+}
+
+// Dequeue from eventQueue, updating graph, populating dirty_queue.
+func (p *Propagator) processEvent() {
+	key, quit := p.eventQueue.Get()
+	if quit {
+		return
+	}
+	defer p.eventQueue.Done(key)
+	event, ok := key.(event)
+	if !ok {
+		glog.Errorf("expect an event, got %v", key)
+		return
+	}
+	obj := event.Obj
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		glog.Errorf("cannot access obj: %v", err)
+		return
+	}
+	typeAccessor, err := meta.TypeAccessor(obj)
+	if err != nil {
+		glog.Errorf("cannot access obj: %v", err)
+		return
+	}
+	glog.V(6).Infof("Propagator process object: %s/%s, namespace %s, name %s, event type %s", typeAccessor.GetAPIVersion(), typeAccessor.GetKind(), accessor.GetNamespace(), accessor.GetName(), event.Type)
+	switch event.Type {
+	case addEvent:
+		// Check if the node already exsits
+		_, ok := p.uidToNode[accessor.GetUID()]
+		if ok {
+			// We can optionally check if the node in the graph is in sync with that in the Store.
+			return
+		}
+		newNode := &node{
+			identity: objectReference{
+				OwnerReference: metatypes.OwnerReference{
+					APIVersion: typeAccessor.GetAPIVersion(),
+					Kind:       typeAccessor.GetKind(),
+					UID:        accessor.GetUID(),
+					Name:       accessor.GetName(),
+				},
+				Namespace: accessor.GetNamespace(),
+			},
+			dependents: make(map[*node]struct{}),
+			owners:     accessor.GetOwnerReferences(),
+		}
+		p.insertNode(newNode)
+		ownerExists := p.addToOwners(newNode)
+		// Push the object to the dirty queue if none of its owners exists in the Graph.
+		// TODO: when handling such an object in the dirtyQueue, the worker should
+		// enqueue an Add event to the eventQueue if the owner exists. This is
+		// to prevent the following case: GC observes the creation of dependent
+		// D, but hasn't observed the creation of Owner O yet; GC adds D to the
+		// dirtyQueue, worker works on D, and O exists in the API server, so GC
+		// doesn't remove D. Then, Propagator sees the creation of O, but no one
+		// will update the graph to reflect that D is the dependent of O, so
+		// when O is deleted, D will not be cascadingly deleted.
+		if !ownerExists && len(newNode.owners) != 0 {
+			p.gc.dirtyQueue.Add(newNode)
+		}
+
+	case updateEvent:
+		node, ok := p.uidToNode[accessor.GetUID()]
+		if !ok {
+			glog.Errorf("received an update for %v, but cannot find the node in the graph", accessor.GetUID())
```

Comment:
@lavalamp I think this really shouldn't happen in the system. Do you have counter example in mind?